### PR TITLE
Use k8s.gcr.io vanity domain for container images

### PR DESCRIPTION
**Pull request by thockin**
Related issue: https://github.com/kubernetes/release/issues/281

```release-note
Use "k8s.gcr.io" for container images rather than "gcr.io/google_containers".  This is just a redirect, for now, so should not impact anyone materially.  

Documentation and tools should all convert to the new name. Users should take note of this in case they see this new name in the system.
```

### CODE REVIEW

**Comment by thockin** [comment]
@ixdy @spxtr Can I get a consult?  Bazel is complaining about failing to pull `@debian-iptables-amd64`  but it works manually:

```
$ docker pull k8s.gcr.io/debian-iptables-amd64:v8
v8: Pulling from debian-iptables-amd64
bfa86312bfc6: Pull complete 
c7da0c8f87db: Pull complete 
Digest: sha256:2e747bc7455b46350d8e57f05c03e109fa306861e7b2a2e8e1cd563932170cf1
Status: Downloaded newer image for k8s.gcr.io/debian-iptables-amd64:v8
```

File: build/root/WORKSPACE [hunk 1]
```diff
@@ -63,16 +63,16 @@ http_file(
 docker_pull(
     name = "debian-iptables-amd64",
     digest = "sha256:2e747bc7455b46350d8e57f05c03e109fa306861e7b2a2e8e1cd563932170cf1",
-    registry = "gcr.io",
-    repository = "google-containers/debian-iptables-amd64",
+    registry = "ks8.gcr.io",
```
Comment (ixdy) [inline build/root/WORKSPACE]: typo: this should be k8s

Comment (thockin) [reply build/root/WORKSPACE]: DOH!

**Comment by chrislovecnm** [comment]
@thockin I think we are going to have to make internal kops changes to get this to work :)  We are creating manifests with gcr.io/google_containers in them.  The test failure is valid.

cc: @justinsb

**Comment by justinsb** [comment]
From the kubelet logs for the kops aws e2e (http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/54174/pull-kubernetes-e2e-kops-aws/54833/artifacts/ec2-34-208-38-61.us-west-2.compute.amazonaws.com/):

```
E1023 05:44:05.298272    3141 pod_workers.go:186] Error syncing pod 660622dfc479756dd91899e3632e1605 ("kube-controller-manager-ip-172-20-37-95.us-west-2.compute.internal_kube-system(660622dfc479756dd91899e3632e1605)"), skipping: failed to "StartContainer" for "kube-controller-manager" with ImagePullBackOff: "Back-off pulling image \"gcr.io/google_containers/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1\""
```

The way this works for CI builds (or any custom builds) is that we `docker load` the docker image from (in this case): `https://storage.googleapis.com/kubernetes-release-pull/ci/pull-kubernetes-e2e-kops-aws/v1.9.0-alpha.1.1512+08d464b74660e1/bin/linux/amd64/kube-controller-manager.tar`

That is currently tagged as `k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1` (in the manifest.json), but we're then looking for `gcr.io/google_containers/...`.

```
[{"Config":"b0d15d3f555741192ed7e32adea9cc94f6d41cfae8fc7b086e80a4c4171541ff.json","RepoTags":["k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1"],"Layers":["b630c879bcd2666364296b51044c1c62f7e89d9391821992be93d78d4e3cd1ba/layer.tar","318e6c912013897e83f7ef89399f7b53c781400837687d84879681854ab8e65d/layer.tar"]}]
```

Pondering the best solution here...

**Comment by thockin** [comment]
We should be able to access it as either name in GCR - this is local tagging?  Can we just tag both names?

**Comment by justinsb** [comment]
Yes, this is tagging in the `docker save` file, which then becomes a local tag.

I think if we tag and export both names in the build that _should_ work:

```
> wget https://storage.googleapis.com/kubernetes-release-pull/ci/pull-kubernetes-e2e-kops-aws/v1.9.0-alpha.1.1512+08d464b74660e1/bin/linux/amd64/kube-controller-manager.tar
> tar xf kube-controller-manager.tar -O manifest.json
[{"Config":"b0d15d3f555741192ed7e32adea9cc94f6d41cfae8fc7b086e80a4c4171541ff.json","RepoTags":["k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1"],"Layers":["b630c879bcd2666364296b51044c1c62f7e89d9391821992be93d78d4e3cd1ba/layer.tar","318e6c912013897e83f7ef89399f7b53c781400837687d84879681854ab8e65d/layer.tar"]}]
> docker load -i kube-controller-manager.tar 
6a749002dd6a: Loading layer [==================================================>] 1.338 MB/1.338 MB
afa65afec971: Loading layer [==================================================>] 131.1 MB/131.1 MB
Loaded image: k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1
> docker tag k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1 gcr.io/altname/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1
> docker save k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1 gcr.io/altname/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1 -o export.tar
> tar xf export.tar -O manifest.json 
[{"Config":"b0d15d3f555741192ed7e32adea9cc94f6d41cfae8fc7b086e80a4c4171541ff.json","RepoTags":["k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1","gcr.io/altname/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1"],"Layers":["92f6836926b231569884ca4fc0052e1a172f7c1b42ca4bd742ed763a9c90b0a2/layer.tar","bb490a0cc591a014b2d39b88a9271ec42641ddbb2096d49a7a294ff9043a26fb/layer.tar"]}]
> docker load -i export.tar 
Loaded image: k8s.gcr.io/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1
Loaded image: gcr.io/altname/kube-controller-manager:v1.9.0-alpha.1.1512_08d464b74660e1
```

We can also change kops to change the prefix for 1.9, though it does mean that the kops job will be broken until we get both changes in.

If double-tagging works, it seems like it's a win in terms of not breaking existing workflows (i.e. other people may also be doing this and expect the old names).

**Comment by chrislovecnm** [comment]
The maybe using the kops assests API till we get the new tags? Not sure that would even work.

**Comment by thockin** [comment]
How to proceed?  Can kops start by double tagging, we get this in, and then finish the conversion?  I don't know where to look.

**Comment by justinsb** [comment]
@thockin I think the k8s build process needs to double-tag, with the old & new names.

Alternatively we can just force-merge, break kops at head for a few hours, and kops can switch to the new names for k8s 1.9 and above.  If you want to do that, let me know so we can tee up the PR on the kops side.

The problem is that the artifact of the CI build is a tar file from docker save, and AFAIK there's no easy way to load that with a particular name.  And the CI builds aren't pushed to a docker registry.

**Comment by chrislovecnm** [comment]
@thockin is the name change backward compatible?  Meaning will I be able to install k8s 1.4 with k8s.gcr.io

**Comment by thockin** [comment]
@chrislovecnm The new name is literally an alias to the old name, underneath.  A kube 1.0 installer should run just fine with the repo as it exists, alongside a brand new install using the new name.  

```
thockin@thockin-glaptop2 kubernetes make-genfiles-cleanup /$ docker images | grep pause

thockin@thockin-glaptop2 kubernetes make-genfiles-cleanup /$ docker pull k8s.gcr.io/pause
Using default tag: latest
latest: Pulling from pause
a3ed95caeb02: Pull complete 
4964c72cd024: Pull complete 
Digest: sha256:a78c2d6208eff9b672de43f880093100050983047b7b0afe0217d3656e1b0d5f
Status: Downloaded newer image for k8s.gcr.io/pause:latest

thockin@thockin-glaptop2 kubernetes make-genfiles-cleanup /$ docker pull gcr.io/google_containers/pause
Using default tag: latest
latest: Pulling from google_containers/pause
Digest: sha256:a78c2d6208eff9b672de43f880093100050983047b7b0afe0217d3656e1b0d5f
Status: Downloaded newer image for gcr.io/google_containers/pause:latest

thockin@thockin-glaptop2 kubernetes make-genfiles-cleanup /$ docker images | grep pause
gcr.io/google_containers/pause        latest                        350b164e7ae1        3 years ago         240kB
k8s.gcr.io/pause                      latest                        350b164e7ae1        3 years ago         240kB
```

@justinsb I have no qualms breaking kops, except that would leave EVERY PR in the queue to fail until kops changes went in.  That's probably not OK.  How hard is this change on the kops end?

Before we can actually commit this, it needs a review and for someone, anyone, to think it through distinct from me.  Just as a sanity check.  I know the idea of ME making a mistake is pretty laughable, but for the sake of optics at least... :)

**Comment by chrislovecnm** [comment]
If it is changed we can make the change in kops master now.

**Comment by ixdy** [comment]
@thockin I believe the issue is that kops side-loads the docker images (from the foo.tar image files in the kubernetes-server tarball) and assumes the loaded images are in the `gcr.io/google_containers` repository. It gets the tag from the images from the `docker_tag` files, but we don't save the repository name anywhere. Since the saved images now use`k8s.gcr.io`, after sideloading docker can't find them. If we were just pulling the images from the registry it'd be no issue, but that's where we are.

I think there's basically three ways to address this:
1. @justinb's suggestion to tag the saved docker images with both repositories (`k8s.gcr.io` and `gcr.io/google_containers`)
2. Stop side-loading images and instead pull them from the registry directly
3. Save the `metadata.json` from the docker image tarballs as a separate release artifact, and make the various cluster startup scripts that sideload images (kops, gce, etc) use this to figure out the correct docker repository/name/tag, rather than the hacky solution we have right now.

I started thinking about the third option in https://github.com/kubernetes/kubernetes/pull/48956#issuecomment-317086476, but didn't (and still don't) have time to work on it. @justinb's suggestion seems the most practical.

**Comment by chrislovecnm** [comment]
@ixdy if the alias is in place we can just make the switch in kops to use it completely. No work arounds, just make the change.

**Comment by chrislovecnm** [comment]
I may have time to do it today and test it

**Comment by ixdy** [comment]
@chrislovecnm does kops sideload or pull from the registry? if we switch kops to using k8s.gcr.io, I worry that'll break everything before this PR, since side-loading a `gcr.io/google_containers` image won't map. (I'm pretty sure the k8s.gcr.io -> gcr.io/google_containers aliasing happens in GCR, not in Docker.)

**Comment by ixdy** [comment]
(I'm mostly worried about devel/CI versions of kubernetes here. I'm guessing kops will pull release versions from the registry if the sideloaded tag doesn't exist.)

**Comment by justinsb** [comment]
@ixdy yes that's correct.  And AIUI the CI images are not currently pushed to a registry, and so we have to pull them from the tar file.  So for suggestion 2 we would have to push CI images to e.g. gcr.io

For suggestion 3, kops can also poke into the metadata.json from the tar file (or parse stdout from docker load) and figure out the loaded image tag.  I think we're happy to do that, but kops may be the canary here - there may be other people side-loading images, and if so suggestion 1 avoids breaking them.    And this is as easy as spawning `tar xf kube-controller-manager.tar -O manifest.json` so while of course I won't object to your uploading manifest.json separately, if we want to go that path it's not a prerequisite.

**Comment by justinsb** [comment]
And confirmed @ixdy that in general kops prefers using the gcr.io images, but can't do that for devel/CI versions which are not pushed to a registry.

**Comment by chrislovecnm** [comment]
We are worried about breaking users that are deploying custom kops versions. I am guessing that we can make the URL change and this will work. But maybe I am wrong and @justinsb is correct.

If we make the domain change kops
Bundle a k8s dev build the same way CI does
Upload that k8s build
Launch a new cluster with that k8s version

I think the above steps is a good regression test

**Comment by ixdy** [comment]
@justinsb we do push CI builds to `gcr.io/kubernetes-ci-images`, though that's only from the CI cross-build job, so it lags `ci/latest` a bit, and it doesn't handle PRs at all (right now).

**Comment by thockin** [comment]
I'm still not understanding - the alias is in place.  Just double tag.  It's just a name?

**Comment by thockin** [comment]
Posisble tagging fix pushed, with TODOs.  If this passes and is approved, I will send something to kubernetes-dev

File: CHANGELOG-1.4.md [hunk 2]
```diff
@@ -177,15 +177,15 @@ filename | sha256 hash
 ### Other notable changes
 
 * kube-apiserver now drops unneeded path information if an older version of Windows kubectl sends it. ([#44586](https://github.com/kubernetes/kubernetes/pull/44586), [@mml](https://github.com/mml))
-* Bump gcr.io/google_containers/glbc from 0.8.0 to 0.9.2. Release notes: [0.9.0](https://github.com/kubernetes/ingress/releases/tag/0.9.0), [0.9.1](https://github.com/kubernetes/ingress/releases/tag/0.9.1), [0.9.2](https://github.com/kubernetes/ingress/releases/tag/0.9.2) ([#43098](https://github.com/kubernetes/kubernetes/pull/43098), [@timstclair](https://github.com/timstclair))
```
Comment (chrislovecnm) [inline CHANGELOG-1.4.md]: Do we really want to revise past history? Personally overkill to me.

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml [hunk 3]
```diff
@@ -45,7 +45,7 @@ spec:
         scheduler.alpha.kubernetes.io/critical-pod: ''
     spec:
       containers:
-        - image: gcr.io/google_containers/heapster-amd64:v1.5.0-beta.0
+        - image: k8s.gcr.io/heapster-amd64:v1.5.0-beta.0
           name: heapster
```
Comment (chrislovecnm) [inline cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: I am not certain how heapster is released, but I am guessing they will need updates in there repo as well?

**Review by chrislovecnm — COMMENTED** [review]
Looks good overall, but I should not be the only set of eyes on this. A couple of questions.

Also kubernetes.io will need updating as well.

Also release notes. Big release notes

File: cluster/addons/cluster-monitoring/google/heapster-controller.yaml (hunk 3 above)
Comment (thockin) [reply cluster/addons/cluster-monitoring/google/heapster-controller.yaml]: Both names will work for forseeable future.

File: CHANGELOG-1.4.md (hunk 2 above)
Comment (thockin) [reply CHANGELOG-1.4.md]: It was auto-generated mostly.  I am ambivalent about revising history, but the old-name links that float around the internet, the better?

**Comment by thockin** [comment]
w00t!  green.  Won't last long.  Now to build approvals. :)

**Comment by smarterclayton** [comment]
From a really high level this looks like a necessary step (with some risk) to getting public ownership of the artifact production pipeline.  I assume risks will be communicated to affected parties (and potentially public blast) ahead of time?

**Comment by jbeda** [comment]
Agree with @smarterclayton. My only wish is that we could go directly to something like `i.k8s.io` instead of still naming gcr explicitly.  I'm assuming that is a limitation on the 'white label' support available with GCR.

**Comment by bgrant0607** [comment]
We definitely need to move off of google_containers.

We need to identify uses of gcr.io/google_containers outside this repo (in addition to kops), such as by add-ons.

For example, the dashboard pushes to google_containers:
https://github.com/kubernetes/dashboard/releases

Also likely the examples moved to the examples repo. kubernetes.io was already mentioned. contrib is an obvious place to check. 

Perhaps we could ask the maintainers of other repos to check them?

A google search found this reference, which will need to be updated:
https://stackoverflow.com/questions/33472741/what-work-does-the-process-in-container-gcr-io-google-containers-pause0-8-0-d

Making this change sooner rather than later seems like a good idea, however.

/approve

**Comment by bgrant0607** [comment]
cc @luxas

**Comment by chrislovecnm** [comment]
@bgrant0607 not just limited to this repository. All of the repositories in kubernetes and kubernetes-incubator as well.

@jbeda no reason we cannot have two aliases. Use gcr.io for now, annoucexwe are going to end of life the gcr.io address by 1.10 and create two now. i.k8s.io just looks odd through ... probably just me

**Comment by luxas** [comment]
Thanks for the ping @bgrant0607 and thanks for the PR @thockin!

(Disclaimer: I didn't read all the msgs in this conversation before due to time shortage; sorry for ignorance if any)

Ok, so this will point to `gcr.io/google_containers` initially, until we've migrated all relevant Kubernetes specific images to some other place, and then we can switch off `gcr.io/google_containers` to whatever other thing we choose, right?

Now that we actually do this (woot!), we really should make sure the backing registry supports manifest lists ([gcr.io doesn't](https://twitter.com/kubernetesonarm/status/922493654429782016)). Either we fix that or we move to something else when this LB is in front.

I agree with @jbeda, could we hide the `gcr.io` part if possible. What if we want to move to something else tomorrow? Also, how is this redirect handled? Seems counter-intuitive to move from something google-specific (`gcr.io/google_containers`) to some other google-specific redirect (k8s.**_gcr.io_**). I mean, I can't probably send a PR against that redirect if I wanted to, right?

Also, should this redirect single-level or multi-level?
With that I mean; should we really restrict ourselves to only one level of hierarchy when we could have more? For instance:
 - `images.k8s.io/core/kube-apiserver`
 - `images.k8s.io/addons/kubernetes-dashboard`
 - `images.k8s.io/incubation/service-catalog-apiserver`

vs. a "flat" `images.k8s.io`.
Also, the image registries the incubator projects are using is kind of broken WRT consistency.

(FWIW; what I wrote above is for illustration, not something I really propose)

And this goes back to permissions => only _some_ Googlers from the k8s team have access to the full-blown `google_containers` (for possible readers of this thread), and this is one of the main motivations for moving away from it. Eventually, having fine-grained permissions for registries (possibly behind a vanity domain like this), will help k8s build releases w/o Google-only creds.

Note: We also have various other domains other than `google_containers` we should consider for this work:
 - kubernetes-e2e-test-images
 - google-samples
 - k8s-authenticated-test
 - k8s-testimages
etc. (probably more)

> Some kind of test that looks for "gcr.io/google_containers" would help prevent new instances from creeping back in.

:+1:
Maybe some kind of metrics for new pushes/pulls to the repo as well so we could actually see the decline in usage

TL;DR;
 - I do think we need a proposal/design doc for this. Not that I like that process more than moving fast, but as one of the major changes in build infra over the lifespan for k8s so far, I think we should think through the scenarios and what we want to achieve before breaking stuff and possibly moving in a direction we don't want later.
 - Let's think through what we need from the backing registry (manifest list support, good ACL mgmt, etc.)
 - Let's think about how to include incubated and/or other "non-core" (in the k8s org still) repo images in this domain, so we don't end up with one image pushing policy per repo.
 - Let's decide what the domain actually should be, and whether it should or shouldn't be Google-controlled.
 - I'm excited to see this happen!

**Comment by dims** [comment]
i like the `images.k8s.io` Not sure about hierarchy, since the images have tags which can be rich enough..

**Comment by thockin** [comment]
Lots of comments:

This is an alias WITHIN gcr.  It is not a blanket alias that we can move to somewhere else.  We could MAYBE do a custom domain with gcr (images.k8s.io) but that is a lot more complicated, and no guarantees that what we would eventually move to (what?) would support custom names.

I guess I am basically saying this is a commitment to GCR.  @luxas it will support manifests.  @jbeda This is why I prototyped that nginx mess, but once I looked at the gcr traffic volume, I decided I *really* don't want to be responsible for the HA on this.  Also, looking at the billing for it, I doubt very much that we're moving ANYWHERE unless it is 100% comped.  The bill is substantial. 

Yes, I will scrub other kube repos if we go with this.  Fortunately, no rush.

A test would be nice.  Someone send me a PR? :)

We can add arbitrary subdirs of this, and we should probably flatten all of our other GCR buckets into this.

Permissions and non-googler access will be more easily fixed once we move to a CNCF-owned project.  I am working on that in parallel, but this has to happen first.

**Comment by thockin** [comment]
Thoughts on my last update?  I'm not going to rebase this again until I have general buy-in.

File: build/build-image/cross/Makefile [hunk 4]
```diff
@@ -21,7 +21,7 @@ TAG=$(shell cat VERSION)
 all: push
 
 build:
-	docker build --pull -t gcr.io/google_containers/$(IMAGE):$(TAG) .
+	docker build --pull -t k8s.gcr.io/$(IMAGE):$(TAG) .
 
 push: build
-	gcloud docker --server=gcr.io -- push gcr.io/google_containers/$(IMAGE):$(TAG)
+	gcloud docker --server=gcr.io -- push k8s.gcr.io/$(IMAGE):$(TAG)
```
Comment (KingEmet) [inline build/build-image/cross/Makefile]: --server=k8s.gcr.io

**Comment by KingEmet** [comment]
2 benefits to k8s.gcr.io (as opposed to alternative aliases):
* TLS termination for free, Google handles cert renewal.
* No additional infrastructure is needed, e.g. @thockin 's nginx proxy.

Using k8s.gcr.io at least allows for the underlying repo (i.e. project) to be changed transparently. As was mentioned, manifest list support is a pending registry feature. There are others relevant to the discussion that I'm not at liberty to disclose, but @thockin is also privy to 😸

File: cluster/images/kubemark/Makefile [hunk 5]
```diff
@@ -15,18 +15,17 @@
...
 
 all: gcloudpush
 
 build:
-	docker build --pull -t $(REGISTRY)/$(PROJECT)/kubemark .
+	docker build --pull -t $(REGISTRY)/kubemark .
 
 gcloudpush: build
-	gcloud docker -- push $(REGISTRY)/$(PROJECT)/kubemark
+	gcloud docker -- push $(REGISTRY)/kubemark
```
Comment (KingEmet) [inline cluster/images/kubemark/Makefile]: For all Makefiles performing pushes to k8s.gcr.io:
sed 's/gcloud docker -- push/gcloud docker --server=k8s.gcr.io -- push/g'
This'll be needed until k8s.gcr.io becomes a default registry for `gcloud docker`

Comment (thockin) [reply cluster/images/kubemark/Makefile]: uggh.  Done.

**Comment by thockin** [comment]
rebased and updated with new users

**Comment by luxas** [comment]
> Yes, I will scrub other kube repos if we go with this

What do you mean with that? Get rid of `gcr.io/k8s-testimages/` etc. ?

> We can add arbitrary subdirs of this, and we should probably flatten all of our other GCR buckets into this

So we can have hierarchy for different things with different ACLs eventually you mean?

> Permissions and non-googler access will be more easily fixed once we move to a CNCF-owned project. I am working on that in parallel, but this has to happen first.

Exactly, so the plan is to move everything to `k8s.gcr.io/<image-name>` first, pointing to `gcr.io/google_containers`, then move the redirect to `gcr.io/cncf-kubernetes` (CNCF-owned) or whatever and optionally add a prefix for things that can/should be treated separately (e.g. `test-infra`-specific images now under `gcr.io/k8s-testimages` could like under `k8s.gcr.io/test-infra/*`, service-catalog stuff under `k8s.gcr.io/service-catalog/` etc.), right?

I understand the concerns of maintaining a HA registry setup of this scale; so using GCR solely makes some sense to me, but it would be nice if we actually could get the manifest list issue (and most importantly; potential other stuff that could pop in the future) prioritized in the case we're going all-in on GCR long-term.

**Comment by thockin** [comment]
>> Yes, I will scrub other kube repos if we go with this
>
> What do you mean with that? Get rid of `gcr.io/k8s-testimages/` etc. ?

THAT is a bigger task, involving permissions and data migration and sync, so no.  I meant just scrubbing for "gcr.io/google_containers".

>> We can add arbitrary subdirs of this, and we should probably flatten all of our other GCR buckets into this
>
> So we can have hierarchy for different things with different ACLs eventually you mean?

yes.

>> Permissions and non-googler access will be more easily fixed once we move to a CNCF-owned project. I am working on that in parallel, but this has to happen first.
>
> Exactly, so the plan is to move everything to `k8s.gcr.io/<image-name>` first, pointing to `gcr.io/google_containers`, then move the redirect to `gcr.io/cncf-kubernetes` (CNCF-owned) or whatever and optionally add a prefix for things that can/should be treated separately (e.g. `test-infra`-specific images now under `gcr.io/k8s-testimages` could like under `k8s.gcr.io/test-infra/*`, service-catalog stuff under `k8s.gcr.io/service-catalog/` etc.), right?

Yes, but those can be done as 2 steps, in either order :)

> I understand the concerns of maintaining a HA registry setup of this scale; so using GCR solely makes some sense to me, but it would be nice if we actually could get the manifest list issue (and most importantly; potential other stuff that could pop in the future) prioritized in the case we're going all-in on GCR long-term.

It's coming.  They are highly incentivized to keep kubernetes happy, but they also have 1000 things to do.

**Comment by chrislovecnm** [comment]
man 1.9 has been flakey - not good

/retest

**Review by luxas — APPROVED** [review]
Are we good to go with this one in order to get it in for v1.9?

I'm very happy with this

**Comment by thockin** [comment]
Trying to get green tests

File: cluster/addons/metadata-proxy/gce/metadata-proxy.yaml [hunk 6]
```diff
@@ -38,7 +38,7 @@ spec:
       dnsPolicy: Default
       containers:
       - name: metadata-proxy
-        image: gcr.io/google_containers/metadata-proxy:v0.1.4
```
Comment (ixdy) [inline cluster/addons/metadata-proxy/gce/metadata-proxy.yaml]: I think this was a bad rebase - metadata proxy now uses a `v` in the tag as of 0.1.4; previously there was no v.

File: build/build-image/cross/Makefile (hunk 4 above)
Comment (thockin) [reply build/build-image/cross/Makefile]: Can you comment on the need for this?  I don't understand this flag

File: cluster/addons/metadata-proxy/gce/metadata-proxy.yaml (hunk 6 above)
Comment (thockin) [reply cluster/addons/metadata-proxy/gce/metadata-proxy.yaml]: good catch

**Comment by thockin** [comment]
I honestly have no idea

/retest pull-kubernetes-e2e-kops-aws

**Comment by thockin** [comment]
rebasing again

**Comment by stp-ip** [comment]
Seeing the value in using k8s.gcr.io for now, but in the long run I would love to see us move to containers.k8s.io or something similar. We are working internally on a implementation on vanity container urls (just doing the redirects and auto TLS) by using DNS TXT records so the various parties don't actually need to host the HA redirect backend, but still control the data and make switching easy.

**Comment by luxas** [comment]
@thockin this needs a rebase as you happened to rebase this in the short timeframe between v1.9.0-beta.0 was cut and https://github.com/kubernetes/kubernetes/commit/a82460d772b94e2b11b65daa47c0a7fb4854b0ad committed, and hence the verify tests are failing...

**Comment by BenTheElder** [comment]
cross failure is #56002

**Comment by luxas** [comment]
Code freeze is today, are we on track to LGTM this beast and get it in?
I think we should do that if only possible

**Comment by thockin** [comment]
I have rebased it at least 50 times.  I'd love to get it in for 1.9, but given the horroshow that is test flakes crossed with rebase after rebase, it's hard to get this green.

**Comment by dims** [comment]
@thockin if you set the milestone to 1.9, add all the labels ... i can baby sit this one

**Comment by thockin** [comment]
I have rebased it 3 times TODAY.  It might have to go in after the freeze actually starts, if it goes for this cycle

**Comment by luxas** [comment]
/retest
unit failure is a flake and the cross build is broken anyway right now :(
I believe that the node-e2e failure is a flake

**Comment by luxas** [comment]
can't tell whether this is a flake or not...
/test pull-kubernetes-e2e-gce

**Comment by KingEmet** [comment]
`gcloud docker` now authenticates k8s.gcr.io by default as of gcloud sdk 182.0.0

**Comment by luxas** [comment]
@thockin What should we do with this one? Is it somehow possible to try to get it in early in the v1.10 timeframe so we have lots of time to notice other possible regressions or side-effects?

I know this is subject to rebase hell; one option would also be to bump priority of this one to the highest once green or temporarily block other merges so this can go in successfully so you don't have to rebase 500 times more. I'd really like us to move forward here, now we also have manifest list support in gcr.io, so we can use `registry/binary:version` everywhere, which will then "redirect" to `registry/binary-arch:version` depending on client platform :+1: (https://twitter.com/estesp/status/940322566937501699)

**Comment by thockin** [comment]
Rebased.  I am going to add priority to it.

I still need an LGTM

**Comment by thockin** [comment]
rebased again

**Comment by luxas** [comment]
Thanks @thockin!

Let's discuss the remaining action items in https://github.com/kubernetes/release/issues/270 everyone.
Also wrote https://twitter.com/kubernetesonarm/status/944170779620626433 for wider visibility, please do the same thing if you can.

**Comment by thockin** [comment]
Working on an updated form of this with what I think will be better global-ness (as best GCR can provide today).  Will need to touch many ancillary repos again. :(

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/pull/57824
